builder/node.go

```diff
@@ -163,6 +166,45 @@ func (b *Builder) LoadNodes(ctx context.Context, withData bool) (_ []Node, err e
 	return b.nodes, nil
 }
 
+func (n *Node) MarshalJSON() ([]byte, error) {
+	var status string
+	if n.DriverInfo != nil {
+		status = n.DriverInfo.Status.String()
+	}
+	var err string
+	if n.Err != nil {
+		status = "error"
+		err = strings.TrimSpace(n.Err.Error())
+	}
+	var pp []string
+	for _, p := range n.Platforms {
+		pp = append(pp, platforms.Format(p))
+	}
+	return json.Marshal(struct {
```

Comment:
I was initially wondering if we should have an actual type for this; possibly if we need specific formatting functions for printing, we could define those on the type.

Then I wondered what we needed the ad-hoc type for (was it because of recursion happening during MarshalJSON?); but then noticed the platforms handling above.

Do you know what specifically was needed to use the `platforms.Format()`? Out of curiousity, I tried what happened if we use `[]ocispecs.Platform` as-is, and (at a glance :warning:), the output seems to look fine.

This is still using the ad-hoc type, but updating it to use OCI platform;

```patch
diff --git a/builder/node.go b/builder/node.go
index 665413f7..2954b153 100644
--- a/builder/node.go
+++ b/builder/node.go
@@ -5,7 +5,6 @@ import (
        "encoding/json"
        "strings"

-       "github.com/containerd/containerd/platforms"
        "github.com/docker/buildx/driver"
        ctxkube "github.com/docker/buildx/driver/kubernetes/context"
        "github.com/docker/buildx/store"
@@ -176,25 +175,21 @@ func (n *Node) MarshalJSON() ([]byte, error) {
                status = "error"
                err = strings.TrimSpace(n.Err.Error())
        }
-       var pp []string
-       for _, p := range n.Platforms {
-               pp = append(pp, platforms.Format(p))
-       }
        return json.Marshal(struct {
                Name        string
                Endpoint    string
-               Platforms   []string          `json:",omitempty"`
-               Flags       []string          `json:",omitempty"`
-               DriverOpts  map[string]string `json:",omitempty"`
-               Files       map[string][]byte `json:",omitempty"`
-               Status      string            `json:",omitempty"`
-               ProxyConfig map[string]string `json:",omitempty"`
-               Version     string            `json:",omitempty"`
-               Err         string            `json:",omitempty"`
+               Platforms   []ocispecs.Platform `json:",omitempty"`
+               Flags       []string            `json:",omitempty"`
+               DriverOpts  map[string]string   `json:",omitempty"`
+               Files       map[string][]byte   `json:",omitempty"`
+               Status      string              `json:",omitempty"`
+               ProxyConfig map[string]string   `json:",omitempty"`
+               Version     string              `json:",omitempty"`
+               Err         string              `json:",omitempty"`
        }{
                Name:        n.Name,
                Endpoint:    n.Endpoint,
-               Platforms:   pp,
+               Platforms:   n.Platforms,
                Flags:       n.Flags,
                DriverOpts:  n.DriverOpts,
                Files:       n.Files,
```

Reply:
With the above, the output _seems_ to be fine (but perhaps there's specific corner-cases we're not accounting for?)

```bash
./bin/build/docker-buildx ls
NAME/NODE           DRIVER/ENDPOINT     STATUS    BUILDKIT                           PLATFORMS
default *           docker
 \_ default          \_ default         running   22.06.0-beta.0-902-g2708be0db4.m   linux/arm64, linux/amd64, linux/amd64/v2, linux/riscv64, linux/ppc64le, linux/s390x,…
desktop-linux       docker
 \_ desktop-linux    \_ desktop-linux   running   22.06.0-beta.0-902-g2708be0db4.m   linux/arm64, linux/amd64, linux/amd64/v2, linux/riscv64, linux/ppc64le, linux/s390x,…
foo                 docker
 \_ foo              \_ foo             error

Failed to get status for foo (foo): Unix socket path "/var/folders/6f/tz5jf4nn1_n5jb0ctrrw5p2w0000gn/T/TestBuildWithBuildercustom_context67217891/001/docker.sock" is too long
```

👆 😂 Ah, fun; looks like I ran some integration test on my machine, and macOS "temp" directories are symlinks, and the _actual_ location looks to be too long.


I also tried outputing only the `.Platforms`, which also seems to work, although (not sure where those come from), it looks like there's empty lines in between each row 🤔 

```bash
 ./bin/build/docker-buildx ls --format '{{.Platforms}}'

linux/arm64, linux/amd64, linux/amd64/v2, linux/riscv64, linux/ppc64le, linux/s390x, linux/mips64le, linux/mips64

linux/arm64, linux/amd64, linux/amd64/v2, linux/riscv64, linux/ppc64le, linux/s390x, linux/mips64le, linux/mips64


```

Reply:
> but perhaps there's specific corner-cases we're not accounting for?

Oh! Perhaps this was for the JSON format? As without it, it would render the platform as a struct? (Wondering if perhaps we should actually keep that 🤔)


```bash
 ./bin/build/docker-buildx ls --format json
{"Name":"default","Driver":"docker","Dynamic":false,"Nodes":[{"Name":"default","Endpoint":"default","Platforms":[{"architecture":"arm64","os":"linux"},{"architecture":"amd64","os":"linux"},{"architecture":"amd64","os":"linux","variant":"v2"},{"architecture":"riscv64","os":"linux"},{"architecture":"ppc64le","os":"linux"},{"architecture":"s390x","os":"linux"},{"architecture":"mips64le","os":"linux"},{"architecture":"mips64","os":"linux"}],"Status":"running","Version":"22.06.0-beta.0-902-g2708be0db4.m"}]}
{"Name":"desktop-linux","Driver":"docker","Dynamic":false,"Nodes":[{"Name":"desktop-linux","Endpoint":"desktop-linux","Platforms":[{"architecture":"arm64","os":"linux"},{"architecture":"amd64","os":"linux"},{"architecture":"amd64","os":"linux","variant":"v2"},{"architecture":"riscv64","os":"linux"},{"architecture":"ppc64le","os":"linux"},{"architecture":"s390x","os":"linux"},{"architecture":"mips64le","os":"linux"},{"architecture":"mips64","os":"linux"}],"Status":"running","Version":"22.06.0-beta.0-902-g2708be0db4.m"}]}
{"Name":"foo","Driver":"docker","Dynamic":false,"Nodes":[{"Name":"foo","Endpoint":"foo","Status":"error","Err":"Unix socket path \"/var/folders/6f/tz5jf4nn1_n5jb0ctrrw5p2w0000gn/T/TestBuildWithBuildercustom_context67217891/001/docker.sock\" is too long"}]}
```

Reply:
Oh, and I guess it doesn't show the `default` / `native` platform; (`linux/arm64*` vs `linux/arm64`); 

edit: ⚠️ actually, that's not it (where is that added?)

If we really want a string presentation in all cases, we could have a local type that wraps `oci.Platform` and implements `MarshalText`;

```patch
diff --git a/builder/node.go b/builder/node.go
index 665413f7..5ee00381 100644
--- a/builder/node.go
+++ b/builder/node.go
@@ -166,6 +166,12 @@ func (b *Builder) LoadNodes(ctx context.Context, withData bool) (_ []Node, err e
        return b.nodes, nil
 }

+type Platform ocispecs.Platform
+
+func (p *Platform) MarshalText() ([]byte, error) {
+       return []byte(platforms.Format(ocispecs.Platform(*p))), nil
+}
+
 func (n *Node) MarshalJSON() ([]byte, error) {
        var status string
        if n.DriverInfo != nil {
@@ -176,14 +182,14 @@ func (n *Node) MarshalJSON() ([]byte, error) {
                status = "error"
                err = strings.TrimSpace(n.Err.Error())
        }
-       var pp []string
+       var pp []Platform
        for _, p := range n.Platforms {
-               pp = append(pp, platforms.Format(p))
+               pp = append(pp, Platform(p))
        }
        return json.Marshal(struct {
                Name        string
                Endpoint    string
-               Platforms   []string          `json:",omitempty"`
+               Platforms   []Platform        `json:",omitempty"`
                Flags       []string          `json:",omitempty"`
                DriverOpts  map[string]string `json:",omitempty"`
                Files       map[string][]byte `json:",omitempty"`
```

Reply:
> Do you know what specifically was needed to use the `platforms.Format()`?

This is for the `--format json` so we are consistent with table / raw result.

> I also tried outputing only the `.Platforms`, which also seems to work, although (not sure where those come from), it looks like there's empty lines in between each row 🤔

Blank lines are the builders which don't carry platforms (only nodes). I'm not sure how we can handle this with custom go template as builders and nodes are really different.

> Oh, and I guess it doesn't show the `default` / `native` platform; (`linux/arm64*` vs `linux/arm64`);

`*` on platform means user enforces this platform when creating the builder or appending a node (`create --platform linux/amd64`) so it will build on this node if it matches the target platform. Atm table, raw format shows supported and preferred platforms but json format doesn't (maybe it should?).